core/corehttp/gateway_handler_unixfs__redirects.go

```diff
@@ -0,0 +1,260 @@
+//
+// Example 3:
+// Another possibility is that the path corresponds to a rewrite rule (i.e. a rule with a status of 200).
+// In this case, we don't perform a redirect, but do need to return a `path.Resolved` and `path.Path` corresponding to
+// the rewrite destination path.
+//
+// Note that for security reasons, redirect rules are only processed when the request has origin isolation.
+// See https://github.com/ipfs/specs/pull/290 for more information.
+func (i *gatewayHandler) serveRedirectsIfPresent(w http.ResponseWriter, r *http.Request, resolvedPath ipath.Resolved, contentPath ipath.Path, logger *zap.SugaredLogger) (newResolvedPath ipath.Resolved, newContentPath ipath.Path, ok bool, hadMatchingRule bool) {
+	redirectsFile := i.getRedirectsFile(r, contentPath, logger)
+	if redirectsFile != nil {
+		redirectRules, err := i.getRedirectRules(r, redirectsFile)
+		if err != nil {
+			internalWebError(w, err)
+			return nil, nil, false, true
+		}
+
+		redirected, newPath, err := i.handleRedirectsFileRules(w, r, contentPath, redirectRules)
+		if err != nil {
+			err = fmt.Errorf("trouble processing _redirects file at %q: %w", redirectsFile.String(), err)
+			internalWebError(w, err)
+			return nil, nil, false, true
+		}
+
+		if redirected {
+			return nil, nil, false, true
+		}
+
+		// 200 is treated as a rewrite, so update the path and continue
+		if newPath != "" {
+			// Reassign contentPath and resolvedPath since the URL was rewritten
+			contentPath = ipath.New(newPath)
+			resolvedPath, err = i.api.ResolvePath(r.Context(), contentPath)
+			if err != nil {
+				internalWebError(w, err)
+				return nil, nil, false, true
+			}
+
+			return resolvedPath, contentPath, true, true
+		}
+	}
+	// No matching rule
+	return resolvedPath, contentPath, false, false
+}
+
+func (i *gatewayHandler) handleRedirectsFileRules(w http.ResponseWriter, r *http.Request, contentPath ipath.Path, redirectRules []redirects.Rule) (bool, string, error) {
+	// Attempt to match a rule to the URL path, and perform the corresponding redirect or rewrite
+	pathParts := strings.Split(contentPath.String(), "/")
+	if len(pathParts) > 3 {
+		// All paths should start with /ipfs/cid/, so get the path after that
+		urlPath := "/" + strings.Join(pathParts[3:], "/")
+		rootPath := strings.Join(pathParts[:3], "/")
+		// Trim off the trailing /
+		urlPath = strings.TrimSuffix(urlPath, "/")
+
+		for _, rule := range redirectRules {
+			// Error right away if the rule is invalid
+			if !rule.MatchAndExpandPlaceholders(urlPath) {
+				continue
+			}
+
+			// We have a match!
+
+			// Rewrite
+			if rule.Status == 200 {
+				// Prepend the rootPath
+				toPath := rootPath + rule.To
+				return false, toPath, nil
+			}
+
+			// Or 400s
+			if rule.Status == 404 {
+				toPath := rootPath + rule.To
+				content404Path := ipath.New(toPath)
+				err := i.serve404(w, r, content404Path)
+				return true, toPath, err
+			}
+
+			if rule.Status == 410 {
```

Comment:
@justincjohnson I missed these before: 
- 410 and 451 are missing sharness tests
- 410 and 451 should behave similarly to what we do for 404, allowing for custom error pages to be rendered informing user why content was taken down.
  - perhaps reuse `i.serve404` – rename it to `i.serve4xx` and make it accept specific code as a parameter?
- `coreiface.ErrResolveFailed` should not be used for them, create dedicated errors

lmk if you have time to address this before Monday, ideally we would include examples of these in the text fixture   that is listed in specs (sadly the CID needs to be updated once again).

Reply:
No problem.  I'll get this done tomorrow.

Reply:
I have made the changes and pushed (not force!).

Reply:
CIDs have been updated in the spec and IPIP as well.